retrieval/discovery/dns/dns.go

```diff
@@ -0,0 +1,234 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dns
+
+import (
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/miekg/dns"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"golang.org/x/net/context"
+
+	"github.com/prometheus/prometheus/config"
+)
+
+const (
+	resolvConf = "/etc/resolv.conf"
+
+	dnsNameLabel = model.MetaLabelPrefix + "dns_name"
+
+	// Constants for instrumentation.
+	namespace = "prometheus"
+)
+
+var (
+	dnsSDLookupsCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "dns_sd_lookups_total",
+			Help:      "The number of DNS-SD lookups.",
+		})
+	dnsSDLookupFailuresCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "dns_sd_lookup_failures_total",
+			Help:      "The number of DNS-SD lookup failures.",
+		})
+)
+
+func init() {
+	prometheus.MustRegister(dnsSDLookupFailuresCount)
+	prometheus.MustRegister(dnsSDLookupsCount)
+}
+
+// Discovery periodically performs DNS-SD requests. It implements
+// the TargetProvider interface.
+type Discovery struct {
+	names []string
+
+	interval time.Duration
+	m        sync.RWMutex
+	port     int
+	qtype    uint16
+}
+
+// NewDiscovery returns a new Discovery which periodically refreshes its targets.
+func NewDiscovery(conf *config.DNSSDConfig) *Discovery {
+	qtype := dns.TypeSRV
+	switch strings.ToUpper(conf.Type) {
+	case "A":
+		qtype = dns.TypeA
+	case "AAAA":
+		qtype = dns.TypeAAAA
+	case "SRV":
+		qtype = dns.TypeSRV
+	}
+	return &Discovery{
+		names:    conf.Names,
+		interval: time.Duration(conf.RefreshInterval),
+		qtype:    qtype,
+		port:     conf.Port,
+	}
+}
+
+// Run implements the TargetProvider interface.
+func (dd *Discovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	defer close(ch)
+
+	ticker := time.NewTicker(dd.interval)
+	defer ticker.Stop()
+
+	// Get an initial set right away.
+	dd.refreshAll(ctx, ch)
+
+	for {
+		select {
+		case <-ticker.C:
+			dd.refreshAll(ctx, ch)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (dd *Discovery) refreshAll(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	var wg sync.WaitGroup
+
+	wg.Add(len(dd.names))
+	for _, name := range dd.names {
+		go func(n string) {
+			if err := dd.refresh(n, ctx, ch); err != nil {
+				log.Errorf("Error refreshing DNS targets: %s", err)
+			}
+			wg.Done()
+		}(name)
+	}
+
+	wg.Wait()
+}
+
+func (dd *Discovery) refresh(name string, ctx context.Context, ch chan<- []*config.TargetGroup) error {
```

Comment:
So far we are passing `ctx` as the first argument consistently.
Let's do this here as well.